pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,283 @@
+	"github.com/aws/karpenter/pkg/apis/config/settings"
+	"github.com/aws/karpenter/pkg/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/cache"
+	interruptionevents "github.com/aws/karpenter/pkg/controllers/interruption/events"
+	"github.com/aws/karpenter/pkg/controllers/interruption/messages"
+	"github.com/aws/karpenter/pkg/controllers/interruption/messages/statechange"
+	"github.com/aws/karpenter/pkg/controllers/providers"
+	"github.com/aws/karpenter/pkg/errors"
+	"github.com/aws/karpenter/pkg/utils"
+
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/operator/scheme"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+func init() {
+	lo.Must0(apis.AddToScheme(scheme.Scheme))
+}
+
+type Action string
+
+const (
+	CordonAndDrain Action = "CordonAndDrain"
+	NoAction       Action = "NoAction"
+)
+
+// Controller is an AWS interruption controller.
+// It continually polls an SQS queue for events from aws.ec2 and aws.health that
+// trigger node health events or node spot interruption/rebalance events.
+type Controller struct {
+	kubeClient                client.Client
+	clk                       clock.Clock
+	recorder                  events.Recorder
+	sqsProvider               *providers.SQS
+	unavailableOfferingsCache *cache.UnavailableOfferings
+	parser                    *EventParser
+}
+
+func NewController(kubeClient client.Client, clk clock.Clock, recorder events.Recorder,
+	sqsProvider *providers.SQS, unavailableOfferingsCache *cache.UnavailableOfferings) *Controller {
+
+	return &Controller{
+		kubeClient:                kubeClient,
+		clk:                       clk,
+		recorder:                  recorder,
+		sqsProvider:               sqsProvider,
+		unavailableOfferingsCache: unavailableOfferingsCache,
+		parser:                    NewEventParser(DefaultParsers...),
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	if !settings.FromContext(ctx).EnableInterruptionHandling {
+		enabled.Set(0)
+		return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+	}
+	queueExists, err := c.sqsProvider.QueueExists(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("checking queue existence, %w", err)
+	}
+	if !queueExists {
+		enabled.Set(0)
+		return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+	}
+	enabled.Set(1)
+	sqsMessages, err := c.sqsProvider.GetSQSMessages(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting messages from queue, %w", err)
+	}
+	if len(sqsMessages) == 0 {
+		return reconcile.Result{}, nil
+	}
+	instanceIDMap, err := c.makeInstanceIDMap(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("making instance id map, %w", err)
+	}
+	errs := make([]error, len(sqsMessages))
+	workqueue.ParallelizeUntil(ctx, 10, len(sqsMessages), func(i int) {
```

Comment:
Worth a peek into lo/parallel to see if there's something more idiomatic.

Reply:
I'd rather overhaul this all at once in a separate PR rather than making a change to this PR at this point.